setup.py

```diff
@@ -1,191 +1,240 @@
-                              "before building.")
-
-        match = re.match(r'.*?\.dev(?P<dev_num>\d+)', full_v)
-        if match is None:
-            dev_num = '0'
-        else:
-            dev_num = match.group('dev_num')
-    else:
-        git_rev = 'Unknown'
-        dev_num = '0'
-
-    if not IS_RELEASED:
-        fullversion += '.dev{0}'.format(dev_num)
-
-    with open(filename, "wt") as fp:
-        fp.write(template.format(version=VERSION,
-                                 full_version=fullversion,
-                                 git_revision=git_rev,
-                                 is_released=IS_RELEASED))
-
-
-def check_python_version():
+
+def _git_output(args):
+    """
+    Call Git with the given arguments and return the output as (Unicode) text.
     """
-    Check that this version of Python is supported.
+    return subprocess.check_output([GIT] + args).decode("utf-8")
 
-    Raise SystemExit for unsupported Python versions.
+
+def _git_info(commit="HEAD"):
     """
-    supported_python_version = (
-        (2, 7) <= sys.version_info < (3,)
-        or (3, 4) <= sys.version_info
-    )
-    if not supported_python_version:
-        sys.exit(
-            (
-                "Python version {0} is not supported by Traits. "
-                "Traits requires Python >= 2.7 or Python >= 3.4."
-            ).format(sys.version_info)
-        )
+    Get information about the given commit from Git.
+
+    Parameters
+    ----------
+    commit : str, optional
+        Commit to provide information for. Defaults to "HEAD".
+
+    Returns
+    -------
+    git_count : int
+        Number of revisions from this commit to the initial commit.
+    git_revision : str
+        Commit hash for HEAD.
+
+    Raises
+    ------
+    EnvironmentError
+        If Git is not available.
+    subprocess.CalledProcessError
+        If Git is available, but the version command fails (most likely
+        because there's no Git repository here).
+    """
+    count_args = ["rev-list", "--count", "--first-parent", commit]
+    git_count = int(_git_output(count_args))
+
+    revision_args = ["rev-list", "--max-count", "1", commit]
+    git_revision = _git_output(revision_args).rstrip()
 
+    return git_count, git_revision
 
-if __name__ == "__main__":
-    check_python_version()
-    write_version_py()
-    from traits import __version__, __requires__
 
-    ctraits = Extension(
-        'traits.ctraits',
-        sources=['traits/ctraits.c'],
-        extra_compile_args=['-DNDEBUG=1', '-O3'],
+def write_version_file(version, git_revision):
+    """
+    Write version information to the version file.
+
+    Overwrites any existing version file.
+
+    Parameters
+    ----------
+    version : packaging.version.Version
+        Package version.
+    git_revision : str
+        The full commit hash for the current Git revision.
+    """
+    with open(VERSION_FILE, "w") as version_file:
```

Comment:
nit - maybe use `io.open`?

Reply:
Yes, I did in a previous version, but setup.py gets picky about Unicode, so I ended up reverting to the "natural" combination of bytes on Python 2 and Unicode on Python 3 (and so plain old "open" everywhere). All the text being read and written is ASCII anyway.

Reply:
For reference: b125c2cd6f2990a90304efdcb21293fb6eca9553

Reply:
Okay, I did end up going back to `io.open`, and adding `u''` prefixes to a handful of important strings. It's an ugly solution: I really want to use `from __future__ import unicode_literals`, but that breaks current setuptools: I get an error resembling the following

```
mirzakhani:traits mdickinson$ python etstool.py update --runtime 2.7
Re-installing in  'traits-test-2.7'
[EXECUTING] edm run -e traits-test-2.7 -- python -m pip install --no-deps .
Processing /Users/mdickinson/Enthought/ETS/traits
    Complete output from command python setup.py egg_info:
    Computed package version: (u'5.2.0.dev1233', u'b2cfb514f7d2674a24bb742b42431194d4a0f37d')
    Writing version to version file /private/var/folders/07/jbbjv8b53bs5y9xyjdyn1zgw0000gn/T/pip-req-build-mZH0B0/traits/version.py.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/07/jbbjv8b53bs5y9xyjdyn1zgw0000gn/T/pip-req-build-mZH0B0/setup.py", line 243, in <module>
        ext_modules=[setuptools.Extension("traits.ctraits", ["traits/ctraits.c"])],
      File "/Users/mdickinson/.edm/envs/traits-test-2.7/lib/python2.7/site-packages/setuptools/extension.py", line 39, in __init__
        _Extension.__init__(self, name, sources, *args, **kw)
      File "/Users/mdickinson/.edm/envs/traits-test-2.7/lib/python2.7/distutils/extension.py", line 106, in __init__
        assert type(name) is StringType, "'name' must be a string"
    AssertionError: 'name' must be a string
```